## keras 방법
- 1) 데이터를 가져온다.
- 2) 데이터를 split한다. x_train, y_train, x_test, y_test
- 3) 그래프를 그린다.
- models.Sequential()
- 4) compile
- 5) fit
- 6) evaluate : test x, y를 넣어주면 정확도를 나타내준다.
- 7) predict : test x를 넣어주면 softmax값을 출력

In [6]:
import keras
from keras import layers
from keras import models
from keras.datasets import mnist
from keras.utils import to_categorical

In [7]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [8]:
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_2 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 64)               

- param #1 : 필터 픽셀 수(3 * 3)과 필터 개수(32) + 32(바이어스 개수) = 320
- param #2 : 3 * 3 * 32 * 64 + 64 = 18496

In [9]:
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
60000/60000 [==============================] - 6s 94us/step - loss: 0.1793 - accuracy: 0.9442
Epoch 2/5
60000/60000 [==============================] - 5s 90us/step - loss: 0.0473 - accuracy: 0.9852
Epoch 3/5
60000/60000 [==============================] - 5s 90us/step - loss: 0.0329 - accuracy: 0.9900
Epoch 4/5
60000/60000 [==============================] - 5s 91us/step - loss: 0.0259 - accuracy: 0.9920
Epoch 5/5
60000/60000 [==============================] - 5s 91us/step - loss: 0.0192 - accuracy: 0.9939


In [11]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 1s 77us/step


In [12]:
test_acc

0.9929999709129333